src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,305 @@
+import json
+import os
+from datetime import datetime, timezone
+from functools import wraps
+from logging import Logger, getLogger
+from typing import Callable, Dict
+
+import boto3
+
+from .enums import StateEnum, WFEventType
+from .eventdb import EventDB
+from .statedb import StateDB
+from .utils import SNSPublisher
+
+
+class WorkflowEventManager:
+    """A class for managing payload state change events, including:
+    1. storage of state (DynamoDB)
+    2. storage of data for workflow metrics (Timestream)
+    3. notifications of Cirrus decisions and/or workflow status changes (SNS).
+
+    Other than `announce`, which is used for announcement of malformed
+    payloads/messages, the public functions here are aimed for use in the `process` and
+    `update-state` lambdas.
+    """
+
+    def __init__(
+        self: "WorkflowEventManager",
+        logger: Logger = None,
+        statedb: StateDB = None,
+        eventdb: EventDB = None,
+        batch_size: int = 10,
+    ):
+        self.logger = logger if logger is not None else getLogger(__name__)
+        self._boto3_session = boto3.Session()
+        wf_event_topic_arn = os.getenv("CIRRUS_WORKFLOW_EVENT_TOPIC_ARN")
+        self.event_publisher = (
+            SNSPublisher(
+                wf_event_topic_arn,
+                logger=self.logger,
+                boto3_session=self._boto3_session,
+                batch_size=batch_size,
+            )
+            if wf_event_topic_arn
+            else None
+        )
+        self.statedb = statedb if statedb is not None else StateDB.get_singleton()
+        self.eventdb = (
+            eventdb if eventdb is not None else EventDB(session=self._boto3_session)
+        )
```

Comment:
I think we don't need the `boto3.Session()` at all on this class?

I get concerned about the `event_publisher` and `eventdb` instances because the client's are being initialized within the scope of a lambda handler. Perhaps though that we aren't seeing issues from this means that lambda bug from the past is no more?

Reply:
😬 The only think I have for this is a email from an AWS support interaction, which I now realize was specifically an issue with a nodejs lambda. I think I saw issues with a python lambda, but now I'm not sure anymore. Maybe we don't see issues with this now because it never affected python. Or maybe boto3 fixed the issue before the nodejs aws-sdk?

I'd say the fact that timestream client hasn't been causing an issue here means I'm just causing more problems pointing out things that aren't a problem. Except I just realized that we _have_ been using a singleton `StateDB`, which has had the reference to the timestream client--so we have gotten a singleton there by composition?

I hesitate to break that pattern here, with us specifically reinstantiating the `WorkflowEventManager` on each lambda execution. This concern also extends to the `SNSPublisher` and `SQSPublisher` classes.

Anyway, the issue was that we were seeing the error `The security token included in the request is invalid`, and it appeared to be intermittently affecting lambda executions when an execution environment was reused to process additional messages. Resolving the problem was as easy as moving our client instantiation out of the `handler`.

I am still frustrated by this issue because I could not get support to explain to me what the difference was between client inside vs outside the handler. Particularly because they said the error message usually means the temporary security credentials granted to the lambda by STS have expired, but IIRC we were seeing the issue manifest in way less than the default hour lifetime those creds get.

🤷 Not sure this helps at all.

Reply:
> I hesitate to break that pattern here, with us specifically reinstantiating the `WorkflowEventManager` on each lambda execution. 

Is that a motion for adopting the singleton pattern for `WorkflowEventManager`?  Let's discuss today.  

> This concern also extends to the `SNSPublisher` and `SQSPublisher` classes.

Seems like this also will need more changes then.